designs/sam_publish_app_cmd.rst

```diff
@@ -0,0 +1,342 @@
+.. contents:: **Table of Contents**
+   :depth: 2
+   :local:
+
+``sam publish app`` command
+====================================
+
+This is the design for a command to publish an application to `AWS Serverless Application Repository (SAR)`_ with a SAM
+template. It can be used to create a new application and its first version, update exisitng application's metadata, create
+a new version of the application, and manage application permissions.
+
+.. _AWS Serverless Application Repository (SAR): https://aws.amazon.com/serverless/serverlessrepo/
+
+
+What is the problem?
+--------------------
+To publish an app to AWS Serverless Application Repository, customers need to go through the following steps: first upload
+the application code and SAM template to an Amazon S3 bucket, correctly set S3 bucket policy that grants the service read
+permissions for artifacts uploaded to S3, then open the AWS Serverless Application Repository console and provide information
+in a bunch of input boxes. If they use the AWS CLI, they need to pass all the information as parameters, and it's easy to make
+a mistake while typing in the command line.
+
+
+What will be changed?
+---------------------
+In this proposal, we will be providing a new command, ``sam publish app``, which takes a SAM template as input and publishes
+an application to AWS Serverless Application Repository using applicaiton metadata specified in the template. Customers
+need to provide application metadata information in the template, then ``sam package`` will handle uploading local files to S3,
+and ``sam publish app`` will create the app in Serverless Application Repository. We will also provide sharing options to set
+application permission policies.
+
+
+Success criteria for the change
+-------------------------------
+#. Support all the following use cases:
+
+   * Create new application and its first version in SAR using ``sam publish app``
```

Comment:
so is sam publish app a constant upsert operation?

Reply:
I dont know if we should have all this interwoven logic all in one command. Thoughts?

Reply:
Yeah it's always a upsert operation. We want to simplify the user experience so that whenever customers have edited their apps locally, they can simply run `sam publish app -t template.yaml` to publish the changes. The command either creates a new app if it doesn't exist, or updates the app with local changes. The interwoven logic is actually hidden in implementation details so that customers don't need to be aware of.